### ISSUE TITLE
Seemingly random i3 freezes (SIGSTOP)

### ISSUE THREAD

**Issue by aksr** [i3/i3#2539]
Arch Linux, 4.12-158-ge48119a

Since September (I think), `i3` randomly freezes.
Sending `SIGCONT` to i3 makes it responsive.
It seems I'm not the only one: https://marc.ttias.be/arch-general/2016-09/msg00067.php

*I'll try to get more info...*

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow http://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by stapelberg** [comment]
My first guess would be that either of the following two things happen:

1. `child.pid` is `-1` in i3bar/src/child.c:stop_child
2. Something else (third-party application) SIGSTOPs i3

**Comment by Airblader** [comment]
How reliably can you reproduce the issue? I agree with Michael's guesses and it'd be easy enough to check the first one if the issue happens reasonably often.

**Comment by Airblader** [comment]
@aksr Any updates?

**Comment by jorge-pessoa** [comment]
I'm also experiencing this issue with Arch 4.8.4-1 and most recent i3 from git. I am not able to consistently reproduce this issue, although I did reduce it's frequency from one freeze every 20 minutes to 1 freeze every 3 or 4 hours by switching from chromium-stable to chrome-beta as browser. I do believe there might be a correlation in that, but I am not entirely sure, as it may have been a coincidence. My i3 logs provide no information on what might be happening. Every input given to i3 while its frozen is executed after the SIGCONT is sent. As I said I can't reproduce this consistently but I would gladly provide whatever log or action you request me to do when i3 freezes.

Thank you for investigating the issue.

**Comment by Tichy** [comment]
I think I had the same problem. It occured while using gtk3 programs such as firefox and transmission with a compositing manager (compton in my case).
For a period I switched to plain i3 without compositing and I hadn't any freeze.

**Comment by jorge-pessoa** [comment]
I can say I experienced freezes while not using any compositing managar so that might not be the end of it

**Comment by aksr** [comment]
Yes, I don't use any compositing manager either.

**Comment by Tichy** [comment]
I had a freeze without compositing and I think that I nailed the cause, at least in my case. I think it was the module window_title_async from py3status or i3ipc-python that it uses. I noticed that the freeze every time occured to me when changing the window focus. Now I'm running i3 with compositing and without that module for a coulde of days and it didn't froze yet.

**Comment by Airblader** [comment]
CC @ultrabug @acrisci

**Comment by stapelberg** [comment]
Interesting. Misbehaving IPC clients can indeed make i3 hang/crash in a variety of ways. We should add that to our debugging steps.

**Comment by joebew42** [comment]
I have the same issue here. I'm on ArchLinux, i3 version 4.13 (2016-11-08). I'm experimenting this issue since september. I'll try to follow your debugging procedure in order to gather more information about this strange behaviour

**Comment by ferreum** [comment]
I've had random freezes of i3 when I started it using dbus-launch like so:

    dbus-launch --exit-with-session -- i3

Apparently dbus tried to read from stdin from the tty that I ran startx from. Since the .xinitrc was run as a background job, the read caused the whole process group to be stopped by SIGTTIN.
The workaround was to redirect stdin from /dev/null like so:

    dbus-launch --exit-with-session -- i3 </dev/null

So maybe it helps to add </dev/null to the script or add the line

    exec </dev/null

to redirect stdin for all processes after it.

**Comment by stapelberg** [comment]
Why is .xinitrc run as a background job in your setup? I was under the impression that .xinitrc is usually run as the session itself, i.e. a foreground job.

Aside from that, I’d recommend to use a display manager instead of startx.

**Comment by Airblader** [comment]
> I’d recommend to use a display manager instead of startx.

That's a recommendation I'd veto… :-)

**Comment by stapelberg** [comment]
Why? Using startx is an additional step you need to perform on every boot, and it requires you lock your text consoles, which many people neglect to do. What’s the advantage?

**Comment by Airblader** [comment]
> Using startx is an additional step you need to perform on every boot

Not if you set up your system to run it automatically upon logging in. :-)

> and it requires you lock your text consoles

Why? All other VTs require a login as well (by default – on Arch, anyway). The only attack would be to somehow crash the X session which would leave you logged in (as user!). But if you can crash that, you can probably also crash i3lock which leaves you logged in despite any display manager anyway.

> What’s the advantage?

I'm not running a software I don't need (and which can cause odd bugs as we've learned in another ticket). :-) I don't use other kinds of sessions and the user login can be handled perfectly fine by the TTY.

I wouldn't necessarily say there's much of an advantage to startx, just that there's not necessarily much advantage to using a display manager either.

**Comment by stapelberg** [comment]
Okay, so I’m not missing any arguments then.

I just really dislike the failure mode of startx: I think when your X session terminates for whatever reason, the resulting state should be a login prompt, not a shell.

Commonly used display managers (e.g. gdm, lightdm, possibly kdm) have the additional advantage that they have a _huge_ user base and hence are more likely to be tested with whichever feature modern desktop environments come up with :). I realize that one can apply this argument to the window manager as well, but I’d say that display managers don’t play a crucial role in the look&feel of the overall user interface (after all, you see them for a few seconds when booting, much like your BIOS/(U)EFI splash screen and boot loader).

If you don’t value either of these reasons, then we’ll just agree to disagree :).

**Comment by ferreum** [comment]
> Why is .xinitrc run as a background job in your setup? I was under the impression that .xinitrc is usually run as the session itself, i.e. a foreground job.

Turns out xinit starts both the X server and .xinitrc in separate process groups, in the background:

    $ ps -H -o stat,pgid,cmd -t /dev/tty2
    STAT  PGID CMD
    Ss+   3919 /bin/sh /usr/bin/startx
    S+    3919   xinit /home/test/.xinitrc -- /etc/X11/xinit/xserverrc :1 vt2 -auth /tmp/serverauth.wgFUTQwZr2
    S<    8813     /usr/lib/xorg-server/Xorg -nolisten tcp :1 vt2 -auth /tmp/serverauth.wgFUTQwZr2
    S     8818     i3

Xorg's and i3's `pgid` are different from xinit's while xinit being the foreground (STAT shows +).

It probably has to do this so it can exit if any of the two processes exit. It then kills the other and exits itself.

Note though that I don't start i3 using dbus-launch anymore. It's not needed anymore because it's now automatically started by the systemd user daemon (which happens by default these days on Arch I think).

> I think when your X session terminates for whatever reason, the resulting state should be a login prompt, not a shell.

Solution to this is starting startx with exec, so it replaces your shell. So you get back to the login prompt if your X session crashes.

**Comment by Airblader** [comment]
> I just really dislike the failure mode of startx: I think when your X session terminates for whatever reason, the resulting state should be a login prompt, not a shell.

Untested, but `startx; exit`? :-) Or what @ferreum said.

> If you don’t value either of these reasons, then we’ll just agree to disagree :).

Just to clarify, from my point of view there's nothing wrong about your recommendation. I just do it differently and thus don't share that recommendation. All in good fun. :-)

**Comment by lasers** [comment]
To rule out between `py3status` or `i3ipc-python` that it uses, I was able to reproduce i3wm freeze using this script `i3ipc-rename-ws.py` with `i3status` so I think the issue is somewhere in `i3ipc-python`.
```
i3ipc-rename-ws.py
-------------------------
#!/usr/bin/env python

import i3ipc

i3 = i3ipc.Connection()

def on_window_focus(i3, e):
    focused = i3.get_tree().find_focused()
    ws_name = "%s:%s" % (focused.workspace().num, focused.window_class)
    i3.command('rename workspace to "%s"' % ws_name)

i3.on("window::focus", on_window_focus)
i3.main()
```

**Comment by JarrColl** [comment]
I have been getting random freezes as well. The last time it happened I was switching tabs in firefox, I am using Compton.

**Comment by ultrabug** [comment]
I'm ashamed to discover the ping only now, sorry @Airblader ...

I'll ping @tobes and @guiniol on this as well since I guess this relates with https://github.com/ultrabug/py3status/issues/941

This is an interesting problem, thanks for the rule out @lasers 👍

**Comment by acrisci** [comment]
I just reproduce the problem simply by running the i3ipc-rename-ws.py script.

**Comment by guiniol** [comment]
Just to be clear because I was cc'd. In the issue I opened (ultrabug/py3status#941), I do not have the whole of i3 freezing, but only py3status, which restarts when I send `SIGCONT`.
py3status freezes sometimes when I lock the screen with i3lock. When I check strace, those are the instances where py3status receives a `SIGSTOP`. When everything works as expected (ie, py3status restarts properly after I unlock the screen), py3status never receives a `SIGSTOP`. I have no idea what is sending those `SIGSTOP` and why they are only sent sometimes and only when I lock the screen.

**Comment by tobes** [comment]
Just to add that in `py3status` we start our output to `i3bar` with `"stop_signal": SIGTSTP` in the header block so we should not be being sent the `SIGSTOP` from my understanding.

I can't trigger things with `i3ipc-rename-ws.py` but maybe a version that is more reproducible might help as maybe a value of `ws_name` causes the issue.

**Comment by greencopper** [comment]
I am running 4.12.8-2-ARCH and are also experiencing random freezes.

At first I thought it was a driver problem and switched graphic card from Intel to Nvidia, but that didn't solve the problem.

I am unable to reproduce the problem as it is completely random. I have however noticed that it occurs mostly while either browsing and scrolling on a webpage in either Firefox or Chromium, or when editing text using Neovim or some other editor and running the cursor across some text.

When it happens I have to do a hard-reboot and I cannot locate any useful information in the systemd journal or elsewhere.

It happens at least once a day when I use my computer and often several times.

**Comment by xstefank** [comment]
Hi, Fedora 26 - i3-4.13-2 - freezes random during maven compilation of big Java project, sometimes waiting several minutes helps, but I've run into freeze for 10+ mins. Today I was logged out into gnome default log in page, but when I logged back in it put me back into my i3 session which was unfrozen

**Comment by orestisf1993** [comment]
For me, when the freeze happens, `ipc_send_message` is stuck in an infinite loop:

https://github.com/i3/i3/blob/7a7481e5784fc4db9b5724c9d2755040f5a47fed/libi3/safewrappers.c#L73-L77

**Comment by orestisf1993** [comment]
I think I found a pattern here.

On a clean i3 session start this script:

```
#!/usr/bin/env python
import i3ipc


def on_ws_focus(self, e):
    print(e.old.name, '->', e.current.name)


for i in range(10):
    try:
        i3 = i3ipc.Connection()
        i3.on('workspace::focus', on_ws_focus)
        i3.main()
    except KeyboardInterrupt:
        print('ctrl-c pressed for connection #', i)
```

each time you ctrl-c it, a new i3ipc connection is started.

Now, run this script:

```
for i in {1..32}
do
i3-msg workspace 2
i3-msg workspace 1
done
```


First time nothing happens, the loop is completed. Now press ctrl-c on the ipc python script and rerun the workspace switch script: i3 freezes.

**Comment by kajzersoze** [comment]
For me it's not random and I can reproduce it whenever I want, I just have to run Lazarus IDE and sometimes right after loading and showing on the screen it locks up, sometimes few minutes after, it's not Lazarus because when I login to different DE it's fine there.

**Comment by Airblader** [comment]
I'm guessing this is another occurence of a misbehaving IPC client. We've discussed this in the past. See #2280.

**Comment by acrisci** [comment]
Try isolating the issue by using the new [go library](https://github.com/i3/go-i3) to test the script or test the script with [sway](https://github.com/SirCmpwn/sway) to try to get some more info where the problem is. If it's really a problem with i3ipc-python, open an issue on my issue tracker and we'll discuss it there.

**Comment by stapelberg** [comment]
I think the problem in the script is that the event IPC connection is not closed upon `KeyboardInterrupt`.

I filed issue #2999 to improve i3’s behavior when IPC clients misbehave.

In the go.i3wm.org/i3 Go package, users must call [Next()](https://godoc.org/go.i3wm.org/i3#EventReceiver.Next) in a loop (i3 might block until the next call to `Next()`) and [Close()](https://godoc.org/go.i3wm.org/i3#EventReceiver.Close) once done. A good way to do this is to use defer to ensure that `Close` is called:

```go
func() {
  recv := i3.Subscribe(i3.WindowEventType)
  defer recv.Close() // ensure cleanup to prevent i3 deadlocks
  for recv.Next() {
    ev := recv.Event().(*i3.WindowEvent)
    // …process ev… — blocking here might mean blocking i3
  }
}()
// At this point, the IPC connection is closed.
```

If receiving all events is not a requirement, one can receive events in a different goroutine and discard events when the processing goroutine is busy:
```go
// Make this channel buffered if required
// (e.g. if each event is persisted to (slow) storage):
windowEvents := make(chan *i3.WindowEvent)
go func() {
  recv := i3.Subscribe(i3.WindowEventType)
  defer recv.Close() // ensure cleanup to prevent i3 deadlocks
  for recv.Next() {
    select {
      case windowEvents <- recv.Event().(*i3.WindowEvent):
      default:
        // discard event to never block i3
    }
  }
}()
for ev := range windowEvents {
  // …process ev…
}
```

**Comment by Crandel** [comment]
I got freezes when I connect Ethernet cable while using wifi. Ubuntu 16.04,  i3 4.14.1-1. Logs show nothing. I use conky instead of i3status

**Comment by dragon788** [comment]
Now that Crandel mentions it I do seem to get occasionally lockups when plugging in a Thunderbolt 3 dock that includes an Ethernet port (USB device internally), and the freezes happen far more often if the Ethernet port is connected than if I am only using wireless. It could be a spurious event from NetworkManager (nm-applet) triggering the issue in my case since I also use that, but interesting coincidence nonetheless.

**Comment by gtp** [comment]
This is old and possibly not directly related, but it's the closest hit to my (workaround solved) problem, so thought it might be useful to somebody. I use i3-input to switch workspaces, and once every couple of days it hangs. Keyboard input is stolen, but mouse still works. Multi-monitor setup, seems to be related to typing too fast, but that's a guess, no peripherals involved.

I tried ssh-ing in from another machine and looking for things to kill, nothing. So for ages I just used to reboot the machine. Yesterday I discovered that if someone Skypes me, which on my machine shows a popup in the top right of the screen, the keyboard unlocks and everything is fine again.

Perhaps that's a clue, or at least useful to someone.
(sorry if this is a thread hijack, I scoured the internet looking for a solution)
